### ISSUE TITLE
[1.5.14] - Clients not being properly closed when connection fails, leading to thread leaks

### ISSUE THREAD

**Issue by mvarela** [protojure/lib#98]
Hi, 
  I've noticed that when the gRPC service I'm trying to use is unavailable, the connection attempts do not fail, but rather hang, and threads are leaked, until eventually the process hangs due to the OS not allowing any more threads to be created. 
Looking at the code, it seems that the issue is in `protojure.internal.grpc.client.providers.http2.jetty/connect`:
 ```
(defn connect [{:keys [host port input-buffer-size idle-timeout ssl] :or {host "localhost" port 80 input-buffer-size 16384 ssl false} :as params}]
  (let [client (HTTP2Client.)
        address (InetSocketAddress. ^String host (int port))
        listener (ServerSessionListener$Adapter.)
        ssl-context-factory (when ssl (SslContextFactory$Client.))]
    (when ssl (.addBean client ssl-context-factory))
    (log/debug "Connecting with parameters: " params)
    (.setInputBufferSize client input-buffer-size)
    (.setInitialStreamRecvWindow client input-buffer-size)
    (when idle-timeout
      (.setIdleTimeout client idle-timeout))
    (.start client)
    (-> (jetty-promise
         (fn [p]
           (.connect client (when ssl ssl-context-factory) address listener p)))
        (p/then (fn [session]
                  (let [context {:client client :session session}]
                    (log/debug "Session established:" context)
                    context)))
        (p/catch (fn [e]
                   (.stop client)
                   (throw e))))))
```

It seems that when the connection fails (e.g., when the service is not available), the call to (.stop client) hangs, and so over time, the threads created by the failed clients leak until no new threads can be created.

I tried adding further logging to that function:

```
(defn connect [{:keys [host port input-buffer-size idle-timeout ssl] :or {host "localhost" port 80 input-buffer-size 16384 ssl false} :as params}]
  (let [client (HTTP2Client.)
        address (InetSocketAddress. ^String host (int port))
        listener (ServerSessionListener$Adapter.)
        ssl-context-factory (when ssl (SslContextFactory$Client.))]
    (when ssl (.addBean client ssl-context-factory))
    (log/debug "Connecting with parameters: " params)
    (.setInputBufferSize client input-buffer-size)
    (.setInitialStreamRecvWindow client input-buffer-size)
    (when idle-timeout
      (.setIdleTimeout client idle-timeout))
    (.start client)
    (-> (jetty-promise
         (fn [p]
           (.connect client (when ssl ssl-context-factory) address listener p)))
        (p/then (fn [session]
                  (let [context {:client client :session session}]
                    (log/debug "Session established:" context)
                    context)))
        (p/catch (fn [e]
                   (log/error "Failed promise")
                   (log/error "stopping client")
                   (.stop client)
                   (log/error (str  "client stopped? " (.isStopped client)))
                   (throw e))))))
```

However, when attempting to connect to a service that's not available, the logs show only:

```
11:00:46.823 [qtp879365180-83] ERROR p.i.g.client.providers.http2.jetty - Failed promise
11:00:46.826 [qtp879365180-83] ERROR p.i.g.client.providers.http2.jetty - stopping client
```

Have you encountered anything similar before? Any ideas of how it could be fixed?

**Comment by ghaskins** [comment]
Hello Martin,

I've not encountered this, though I also do not know if I have tested what happens when the backend is not available.  I did take a quick look at jetty, wondering if the (.start) may possibly return an error that indicates some error state in which (.stop) may not be valid.   However, it seems it returns void, so that doesnt look promising.

One thing I am wondering: are you sure that (.stop) hangs?  Could it be that it is throwing an exception instead?

Try adding a (try (catch)) around the (.stop) and see if that illuminates something.  I will try to poke around ASAP and see if I can fashion a solution as well.

**Comment by mvarela** [comment]
Hi Gregory, thanks for your reply. This is indeed weird... I added the try/catch, and still got the same results when running it normally, however, if I step it through in the CIDER debugger, it actually works.
The connect code reads:
```
(defn connect [{:keys [host port input-buffer-size idle-timeout ssl] :or {host "localhost" port 80 input-buffer-size 16384 ssl false} :as params}]
  (let [client (HTTP2Client.)
        address (InetSocketAddress. ^String host (int port))
        listener (ServerSessionListener$Adapter.)
        ssl-context-factory (when ssl (SslContextFactory$Client.))]
    (when ssl (.addBean client ssl-context-factory))
    (log/debug "Connecting with parameters: " params)
    (.setInputBufferSize client input-buffer-size)
    (.setInitialStreamRecvWindow client input-buffer-size)
    (when idle-timeout
      (.setIdleTimeout client idle-timeout))
    (.start client)
    (-> (jetty-promise
         (fn [p]
           (.connect client (when ssl ssl-context-factory) address listener p)))
        (p/then (fn [session]
                  (let [context {:client client :session session}]
                    (log/debug "Session established:" context)
                    context)))
        (p/catch (fn [e]
                   (log/error "Failed promise")
                   (log/error "stopping client")
                   (log/error (.getMessage e))
                   (try
                     (.stop client)
                     (catch Exception x (log/error (str "Calling .stop threw this " (.getMessage e)))))
                   (log/error (str  "client stopped? " (.isStopped client)))
                   (throw e))))))
```

Alternating normal calls with debugger step-throughs, this is the output:
```


08:34:02.748 [nREPL-session-c11c1358-af49-4d72-bcf2-c58513309c4b] ERROR p.i.g.client.providers.http2.jetty - Failed promise
08:34:03.278 [nREPL-session-c11c1358-af49-4d72-bcf2-c58513309c4b] ERROR p.i.g.client.providers.http2.jetty - stopping client
08:34:05.482 [nREPL-session-c11c1358-af49-4d72-bcf2-c58513309c4b] ERROR p.i.g.client.providers.http2.jetty - Connection refused
08:34:12.847 [nREPL-session-c11c1358-af49-4d72-bcf2-c58513309c4b] ERROR p.i.g.client.providers.http2.jetty - client stopped? true
08:34:37.306 [qtp637257792-100] ERROR p.i.g.client.providers.http2.jetty - Failed promise
08:34:37.306 [qtp637257792-100] ERROR p.i.g.client.providers.http2.jetty - stopping client
08:34:37.307 [qtp637257792-100] ERROR p.i.g.client.providers.http2.jetty - Connection refused
08:35:46.608 [qtp528349316-116] ERROR p.i.g.client.providers.http2.jetty - Failed promise
08:35:46.608 [qtp528349316-116] ERROR p.i.g.client.providers.http2.jetty - stopping client
08:35:46.608 [qtp528349316-116] ERROR p.i.g.client.providers.http2.jetty - Connection refused
08:36:45.398 [nREPL-session-c11c1358-af49-4d72-bcf2-c58513309c4b] ERROR p.i.g.client.providers.http2.jetty - Failed promise
08:36:46.472 [nREPL-session-c11c1358-af49-4d72-bcf2-c58513309c4b] ERROR p.i.g.client.providers.http2.jetty - stopping client
08:36:48.298 [nREPL-session-c11c1358-af49-4d72-bcf2-c58513309c4b] ERROR p.i.g.client.providers.http2.jetty - Connection refused
08:36:56.865 [nREPL-session-c11c1358-af49-4d72-bcf2-c58513309c4b] ERROR p.i.g.client.providers.http2.jetty - client stopped? true
08:37:34.348 [qtp454600783-147] ERROR p.i.g.client.providers.http2.jetty - Failed promise
08:37:34.348 [qtp454600783-147] ERROR p.i.g.client.providers.http2.jetty - stopping client
08:37:34.349 [qtp454600783-147] ERROR p.i.g.client.providers.http2.jetty - Connection refused
```

As you can see, the `client stopped?` bit only appears in the debug runs, for some reason. 

When running it normally, I get a timeout when `deref`ing the client, and the threads leak.
```
(def client (http2/connect {:uri "http://example.com:30001"  :idle-timeout 1}))

(deref client 1000 :timeout);; => :timeout
```

When doing the `deref` after the debug run, I get a `ConnectException`, as expected...

It's quite perplexing. I'll try to do a standalone repro, and see how that goes.

**Comment by mvarela** [comment]
So, I did a standalone repro, and it works as it should. I did notice that I needed to add an explicit dependency on `gclib`, which I don't have on my actual project. This led me to think that maybe there was some issue with dependency versions brought by other packages, so I moved protojure & co up in my deps.edn dependencies list, and now the threads don't get stuck when the connection fails (at least when trying manually). I checked in VisualVM, and they do disappear after a few seconds.

So, I'm not sure what the problem was exactly, but I suspect it was related to some dependency (probably jetty-related, since I'm using ring-jetty too). I'll close this for now, since it doesn't reproduce stand-alone in any case.

Thanks for the help!
